modules/diskq/qdisk.c

```diff
@@ -519,39 +519,38 @@ _read_record_from_disk(QDisk *self, GString *record, guint32 record_length)
   return TRUE;
 }
 
-static inline gint64
-_calculate_new_read_head_position(QDisk *self, guint32 record_length)
+static inline void
+_maybe_apply_non_reliable_corrections(QDisk *self)
 {
-  gint64 new_read_head_position = self->hdr->read_head + record_length + sizeof(record_length);
-
-  if (new_read_head_position > self->hdr->write_head)
-    new_read_head_position = _correct_position_if_max_size_is_reached(self, &new_read_head_position);
+  if (self->options->reliable)
+    return;
 
-  return new_read_head_position;
+  qdisk_empty_backlog(self);
+  g_assert(self->hdr->backlog_len == 0);
+  if (!self->options->read_only)
+    qdisk_reset_file_if_empty(self);
 }
 
+
 static inline void
-_update_positions_after_read(QDisk *self, guint32 record_length)
+_update_position_after_read(QDisk *self, guint32 record_length, gint64 *position)
 {
-  self->hdr->read_head = _calculate_new_read_head_position(self, record_length);
-  self->hdr->length--;
+  gint64 new_read_head_position = *position + record_length + sizeof(record_length);
 
-  if (!self->options->reliable)
-    {
-      self->hdr->backlog_head = self->hdr->read_head;
+  if (new_read_head_position > self->hdr->write_head)
+    new_read_head_position = _correct_position_if_max_size_is_reached(self, new_read_head_position);
 
-      g_assert(self->hdr->backlog_len == 0);
-      if (!self->options->read_only)
-        {
-          qdisk_reset_file_if_empty(self);
-        }
-    }
+  *position = new_read_head_position;
 }
 
 gint64
-qdisk_get_head_position(QDisk *self)
+qdisk_get_next_head_position(QDisk *self)
```

Comment:
Optional: 
for me, I don't get it it will return the value of the next read head, I think it would be better to name the `qdisk_get_next_read_head_position`

Reply:
This naming convention already exists in the QDisk interface, if you check its header file.
Considering QDisk as a ring-buffer on disk, or a LogQueue implementation, "reader" and "writer"  are redundant terms.

Queues usually have a head or tail, which can internally mean they have a reader or writer pointers.
Attila decided not to publish these names, I like the idea.

Note: The getters at the bottom of the QDisk headers are not meant to be public, they are only there so we can do nasty things in unit tests (which we should not do), but those are not part of the real interface.

Reply:
ohh okay, thanks!